skimage/restoration/inpaint.py

```diff
@@ -1,79 +1,200 @@
+    out_flat = np.ascontiguousarray(out.reshape((-1, n_channels)))
+    idx_known = 0
+    idx_unknown = 0
+    mask_pt_n = -1
+    for mask_pt_n, mask_pt_idx in enumerate(zip(*boundary_pts)):
         # Get bounded neighborhood of selected radius
-        b_lo, b_hi = _get_neighborhood(mask_pt_idx, 2, out.shape)
+        b_lo, b_hi = _get_neighborhood(mask_pt_idx, radius, out.shape)
 
-        # Create biharmonic coefficients ndarray
-        neigh_coef = np.zeros(b_hi - b_lo)
-        neigh_coef[tuple(mask_pt_idx - b_lo)] = 1
-        neigh_coef = laplace(laplace(neigh_coef))
+        # Create (truncated) biharmonic coefficients ndarray
+        coef_shape = tuple(hi - lo for lo, hi in zip(b_lo, b_hi))
+        coef_center = tuple(p - lo for p, lo in zip(mask_pt_idx, b_lo))
+        neigh_coef, coef_idx, coefs = _get_neigh_coef(coef_shape,
+                                                      coef_center,
+                                                      coef_cache,
+                                                      dtype=out.dtype)
+
+        # compute corresponding 1d indices into the mask
+        coef_idx = [c + b for c, b in zip(coef_idx, b_lo)]
+        index1d = np.ravel_multi_index(coef_idx, mask.shape)
 
         # Iterate over masked point's neighborhood
-        it_inner = np.nditer(neigh_coef, flags=['multi_index'])
-        for coef in it_inner:
-            if coef == 0:
-                continue
-            tmp_pt_idx = np.add(b_lo, it_inner.multi_index)
-            tmp_pt_i = np.ravel_multi_index(tmp_pt_idx, mask.shape)
-
-            if mask[tuple(tmp_pt_idx)]:
-                matrix_unknown[mask_pt_n, tmp_pt_i] = coef
+        nvals = 0
+        for coef, i in zip(coefs, index1d):
+            if mask_flat[i]:
+                row_idx_unknown[idx_unknown] = mask_pt_n
+                col_idx_unknown[idx_unknown] = i
+                data_unknown[idx_unknown] = coef
+                idx_unknown += 1
             else:
-                matrix_known[mask_pt_n, tmp_pt_i] = coef
-
-    # Prepare diagonal matrix
-    flat_diag_image = sparse.dia_matrix((out.flatten(), np.array([0])),
-                                        shape=(out.size, out.size))
-
-    # Calculate right hand side as a sum of known matrix's columns
-    matrix_known = matrix_known.tocsr()
-    rhs = -(matrix_known * flat_diag_image).sum(axis=1)
+                data_known[idx_known, :] -= coef * out_flat[i, :]
+                nvals += 1
+        if nvals:
+            row_idx_known[idx_known] = mask_pt_n
+            idx_known += 1
+
+    # Call an efficient Cython-based implementation for all interior points
+    row_start = mask_pt_n + 1
+    known_start_idx = idx_known
+    unknown_start_idx = idx_unknown
+    nnz_rhs = _build_matrix_inner(
+        # starting indices
+        row_start, known_start_idx, unknown_start_idx,
+        # input arrays
+        center_i, raveled_offsets, coef_vals, mask_flat.astype(np.int16),
```

Comment:
We can save some memory here if we use `np.uint8` and `char` instead of `short` in `_build_matrix_inner` signature
```suggestion
        center_i, raveled_offsets, coef_vals, mask_flat.astype(np.uint8),
```